Postgres should revoke permissions upon revocation

The Postgres backend drops the created role during revocation, but open sessions are still able to perform queries.

The MySQL backend explicitly revokes permissions first, then drops the user. Open sessions are unable to perform further queries.

The Postgres backend should behave the same way, to ensure that a session that remains open past expiration (for instance, due to a long-running query) is functionally useless.

And we just ran into the following:

~~~
2016/01/05 02:45:43 [INFO] core: pre-seal teardown complete
panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xb code=0x1 addr=0x0 pc=0x8033a0]

goroutine 261549 [running]:
database/sql.(*DB).prepare(0xc8203f1a40, 0xc82135d200, 0x8a, 0xc821355c01, 0x8a, 0x0, 0x0)
        /usr/local/go/src/database/sql/sql.go:871 +0x470
database/sql.(*DB).Prepare(0xc8203f1a40, 0xc82135d200, 0x8a, 0x1, 0x0, 0x0)
        /usr/local/go/src/database/sql/sql.go:849 +0x76
github.com/hashicorp/vault/builtin/logical/postgresql.(*backend).secretCredsRevoke(0xc8206b0480, 0xc8216aea80, 0xc82033a230, 0x0, 0x0, 0x0)
        /gopath/src/github.com/hashicorp/vault/builtin/logical/postgresql/secret_creds.go:110 +0x37b
github.com/hashicorp/vault/builtin/logical/postgresql.(*backend).(github.com/hashicorp/vault/builtin/logical/postgresql.secretCredsRevoke)-fm(0xc8216aea80, 0xc82033a230, 0x5, 0x0, 0x0)
        /gopath/src/github.com/hashicorp/vault/builtin/logical/postgresql/secret_creds.go:33 +0x42
github.com/hashicorp/vault/logical/framework.(*Secret).HandleRevoke(0xc820469f00, 0xc8216aea80, 0x1210e98, 0x0, 0x0)
        /gopath/src/github.com/hashicorp/vault/logical/framework/secret.go:87 +0xac
github.com/hashicorp/vault/logical/framework.(*Backend).handleRevokeRenew(0xc82029cbb0, 0xc8216aea80, 0x1210e98, 0x0, 0x0)
        /gopath/src/github.com/hashicorp/vault/logical/framework/backend.go:333 +0x45c
github.com/hashicorp/vault/logical/framework.(*Backend).HandleRequest(0xc82029cbb0, 0xc8216aea80, 0xc8216aea80, 0x0, 0x0)
        /gopath/src/github.com/hashicorp/vault/logical/framework/backend.go:88 +0x113
github.com/hashicorp/vault/vault.(*Router).Route(0xc822051350, 0xc8216aea80, 0x0, 0x0, 0x0)
        /gopath/src/github.com/hashicorp/vault/vault/router.go:256 +0xb0e
github.com/hashicorp/vault/vault.(*ExpirationManager).revokeEntry(0xc8203a6f00, 0xc8216aea00, 0x0, 0x0)
        /gopath/src/github.com/hashicorp/vault/vault/expiration.go:490 +0x29f
github.com/hashicorp/vault/vault.(*ExpirationManager).Revoke(0xc8203a6f00, 0xc82030f490, 0x63, 0x0, 0x0)
        /gopath/src/github.com/hashicorp/vault/vault/expiration.go:188 +0x1bf
github.com/hashicorp/vault/vault.(*ExpirationManager).expireID(0xc8203a6f00, 0xc82030f490, 0x63)
        /gopath/src/github.com/hashicorp/vault/vault/expiration.go:465 +0x104
github.com/hashicorp/vault/vault.(*ExpirationManager).Restore.func1()
        /gopath/src/github.com/hashicorp/vault/vault/expiration.go:151 +0x37
created by time.goFunc
        /usr/local/go/src/time/sleep.go:129 +0x3a
~~~

--- a/builtin/logical/postgresql/secret_creds.go
+++ b/builtin/logical/postgresql/secret_creds.go
@@ -1,6 +1,7 @@
 package postgresql
 
 import (
+	"database/sql"
 	"fmt"
 	"time"
 
@@ -130,6 +131,32 @@ func (b *backend) secretCredsRevoke(
 		revocationStmts = append(revocationStmts, fmt.Sprintf(
 			"REVOKE ALL PRIVILEGES ON ALL TABLES IN SCHEMA %s FROM %s;",
 			schema, pq.QuoteIdentifier(username)))
+
+		revocationStmts = append(revocationStmts, fmt.Sprintf(
+			"REVOKE USAGE ON SCHEMA %s FROM %s;",
+			schema, pq.QuoteIdentifier(username)))
+	}
+
+	// for good measure, revoke all privileges and usage on schema public
+	revocationStmts = append(revocationStmts, fmt.Sprintf(
+		"REVOKE ALL PRIVILEGES ON ALL TABLES IN SCHEMA public FROM %s;",
+		pq.QuoteIdentifier(username)))
+
+	revocationStmts = append(revocationStmts, fmt.Sprintf(
+		"REVOKE USAGE ON SCHEMA public FROM %s;",
+		pq.QuoteIdentifier(username)))
+
+	// get the current database name so we can issue a REVOKE CONNECT for
+	// this username
+	var dbname sql.NullString
+	if err := db.QueryRow("SELECT current_database();").Scan(&dbname); err != nil {
+		return nil, err
+	}
+
+	if dbname.Valid {
+		revocationStmts = append(revocationStmts, fmt.Sprintf(
+			"REVOKE CONNECT ON DATABASE %s FROM %s;",
+			dbname.String, pq.QuoteIdentifier(username)))
 	}
 
 	// again, here, we do not stop on error, as we want to remove as

